test/src/Model/BrushTest.cpp

```diff
@@ -954,6 +954,267 @@ namespace TrenchBroom {
+        }
+        
+        static void checkCanNotMoveFace(const Brush* brush, const BrushFace* topFace, const Vec3 delta) {
+            const BBox3 worldBounds(4096.0);
+            
+            ASSERT_NE(nullptr, topFace);
+
+            ASSERT_FALSE(brush->canMoveFaces(worldBounds, Polygon3::List { topFace->polygon() }, delta));
+        }
+        
+        static void checkCanMoveTopFace(const Brush* brush, const Vec3 delta) {
+            checkCanMoveFace(brush, brush->findFace(Vec3::PosZ), delta);
+        }
+        
+        static void checkCanNotMoveTopFace(const Brush* brush, const Vec3 delta) {
+            checkCanNotMoveFace(brush, brush->findFace(Vec3::PosZ), delta);
+        }
+        
+        static void checkCanNotMoveTopFaceBeyond127UnitsDown(Brush *brush) {
+            checkCanMoveTopFace(brush, Vec3(0, 0, -127));
+            checkCanNotMoveTopFace(brush, Vec3(0, 0, -128));
+            checkCanNotMoveTopFace(brush, Vec3(0, 0, -129));
+            
+            checkCanMoveTopFace(brush, Vec3(256, 0, -127));
+            checkCanNotMoveTopFace(brush, Vec3(256, 0, -128));
+            checkCanNotMoveTopFace(brush, Vec3(256, 0, -129));
+        }
+        
+        static void checkCanMoveVertex(const Brush* brush, const Vec3 vertexPosition, const Vec3 delta) {
+            const BBox3 worldBounds(4096.0);
+
+            ASSERT_TRUE(brush->canMoveVertices(worldBounds, Vec3::List { vertexPosition }, delta));
+            
+            Brush* brushClone = brush->clone(worldBounds);
+            const Vec3::List movedVertexPositions = brushClone->moveVertices(worldBounds, Vec3::List { vertexPosition }, delta);
+            
+            ASSERT_EQ(Vec3::List { vertexPosition + delta}, movedVertexPositions);
+            
+            delete brushClone;
+        }
+        
+        static void checkMovingVertexDeletes(const Brush* brush, const Vec3 vertexPosition, const Vec3 delta) {
+            const BBox3 worldBounds(4096.0);
+            
+            ASSERT_TRUE(brush->canMoveVertices(worldBounds, Vec3::List { vertexPosition }, delta));
+            
+            Brush* brushClone = brush->clone(worldBounds);
+            const Vec3::List movedVertexPositions = brushClone->moveVertices(worldBounds, Vec3::List { vertexPosition }, delta);
+            
+            ASSERT_EQ(Vec3::List(), movedVertexPositions);
+            
+            delete brushClone;
+        }
+        
+        static void checkCanNotMoveVertex(const Brush* brush, const Vec3 vertexPosition, const Vec3 delta) {
+            const BBox3 worldBounds(4096.0);
+            ASSERT_FALSE(brush->canMoveVertices(worldBounds, Vec3::List { vertexPosition }, delta));
+        }
+        
+        // "Move point" tests
+        
+        // NOTE: Different than movePolygonRemainingPoint, because in this case we allow
+        // point moves that flip the normal of the remaining polygon
+        TEST(BrushTest, movePointRemainingPolygon) {
```

Comment:
It seems to work fine in the editor; I can add some extra assertions here to check the face normal and the order of the vertices